cmd/gardener-node-agent/app/app.go

```diff
@@ -0,0 +1,227 @@
+			return fmt.Errorf("failed fetching access token via bootstrap token: %w", err)
+		}
+	}
+
+	var extraHandlers map[string]http.Handler
+	if cfg.Debugging != nil && cfg.Debugging.EnableProfiling {
+		extraHandlers = routes.ProfilingHandlers
+		if cfg.Debugging.EnableContentionProfiling {
+			goruntime.SetBlockProfileRate(1)
+		}
+	}
+
+	log.Info("Setting up manager")
+	mgr, err := manager.New(restConfig, manager.Options{
+		Logger:                  log,
+		Scheme:                  kubernetes.SeedScheme,
+		GracefulShutdownTimeout: pointer.Duration(5 * time.Second),
+
+		HealthProbeBindAddress: net.JoinHostPort(cfg.Server.HealthProbes.BindAddress, strconv.Itoa(cfg.Server.HealthProbes.Port)),
+		Metrics: metricsserver.Options{
+			BindAddress:   net.JoinHostPort(cfg.Server.Metrics.BindAddress, strconv.Itoa(cfg.Server.Metrics.Port)),
+			ExtraHandlers: extraHandlers,
+		},
+
+		Cache: cache.Options{
+			ByObject: map[client.Object]cache.ByObject{
+				&corev1.Secret{}: {
+					Namespaces: map[string]cache.Config{metav1.NamespaceSystem: {}},
+				},
+			},
+		},
+		LeaderElection: false,
+		Controller: controllerconfig.Controller{
+			RecoverPanic: pointer.Bool(true),
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	log.Info("Setting up health check endpoints")
+	if err := mgr.AddHealthzCheck("ping", healthz.Ping); err != nil {
+		return err
+	}
+	if err := mgr.AddReadyzCheck("informer-sync", gardenerhealthz.NewCacheSyncHealthz(mgr.GetCache())); err != nil {
+		return err
+	}
+
+	log.Info("Adding controllers to manager")
+	if err := controller.AddToManager(mgr, cfg); err != nil {
+		return fmt.Errorf("failed adding controllers to manager: %w", err)
+	}
+
+	log.Info("Starting manager")
+	return mgr.Start(ctx)
+}
+
+func fetchAccessTokenViaBootstrapToken(ctx context.Context, log logr.Logger, restConfig *rest.Config, cfg *config.NodeAgentConfiguration) error {
+	c, err := client.New(restConfig, client.Options{})
+	if err != nil {
+		return fmt.Errorf("unable to create client with bootstrap token: %w", err)
+	}
+
+	secret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: cfg.AccessTokenSecretName, Namespace: metav1.NamespaceSystem}}
+	log.Info("Fetching access token secret", "secret", client.ObjectKeyFromObject(secret))
+	if err := c.Get(ctx, client.ObjectKeyFromObject(secret), secret); err != nil {
+		return fmt.Errorf("failed fetching access token from API server: %w", err)
+	}
+
+	token := secret.Data[resourcesv1alpha1.DataKeyToken]
+	if len(token) == 0 {
+		return fmt.Errorf("secret key %q does not exist or empty", resourcesv1alpha1.DataKeyToken)
+	}
+
+	log.Info("Successfully fetched access token from API server, writing it to disk", "path", nodeagentv1alpha1.TokenFilePath)
+	if err := os.WriteFile(nodeagentv1alpha1.TokenFilePath, token, 0600); err != nil {
+		return fmt.Errorf("unable to write access token to %s: %w", nodeagentv1alpha1.TokenFilePath, err)
+	}
+	log.Info("Token written to disk")
```

Comment:
We could move the token write part to the end of the function just before writing the kubeconfig. Then, we would not create an orphaned token on the file system if encoding of the kubeconfig fails.